lib/Target/SimFHE/SimFHEEmitter.cpp

```diff
@@ -0,0 +1,179 @@
+    for (auto arg : func.getArguments()) {
+      if (isa<lwe::NewLWECiphertextType>(arg.getType()))
+        os << "scheme_params.fresh_ctxt,";
+      else if (isa<lwe::NewLWEPlaintextType>(arg.getType()))
+        os << "params.PolyContext(scheme_params.fresh_ctxt.logq, "
+              "scheme_params.fresh_ctxt.logN, "
+              "scheme_params.fresh_ctxt.dnum,1), ";
+      else
+        return func.emitOpError("Unsupported argument type for SimFHEEmitter: ")
+               << arg.getType();
+    }
+    os << "scheme_params]))\n";
+  }
+  os.unindent();  // exit internal for loop
+  os << kMainPostlude << "\n";
+  return success();
+}
+
+LogicalResult SimFHEEmitter::printOperation(func::FuncOp funcOp) {
+  os << "def " << funcOp.getName() << "(";
+  os << heir::commaSeparatedValues(funcOp.getArguments(),
+                                   [&](Value value) { return getName(value); });
+  if (funcOp.getNumArguments() > 0) os << ", ";
+  os << "scheme_params : params.SchemeParams):\n";
+  os.indent();
+  os << "stats = PerfCounter()\n";
+  for (Block &block : funcOp.getBlocks()) {
+    for (Operation &op : block.getOperations()) {
+      if (failed(translate(op))) {
+        return failure();
+      }
+    }
+  }
+  os << "return stats\n";
+  os.unindent();
+  os << "\n";
+  return success();
+}
+
+LogicalResult SimFHEEmitter::printOperation(func::ReturnOp op) {
+  // return is handled via accumulating stats; nothing to do
+  return success();
+}
+
+#define UNARY_EMIT(OPTYPE, GETTER, FUNCNAME)                 \
+  LogicalResult SimFHEEmitter::printOperation(OPTYPE op) {   \
+    std::string input = getName(op.GETTER());                \
+    os << "stats += evaluator." FUNCNAME "(" << input        \
+       << ", scheme_params.arch_param)\n";                   \
+    os << getName(op.getResult()) << " = " << input << "\n"; \
+    return success();                                        \
+  }
+
+#define UNARY_EMIT_DROP(OPTYPE, GETTER, FUNCNAME)                   \
+  LogicalResult SimFHEEmitter::printOperation(OPTYPE op) {          \
+    std::string input = getName(op.GETTER());                       \
+    os << "stats += evaluator." FUNCNAME "(" << input               \
+       << ", scheme_params.arch_param)\n";                          \
+    os << getName(op.getResult()) << " = " << input << ".drop()\n"; \
+    return success();                                               \
+  }
+
+#define BINARY_EMIT(OPTYPE, GETTER, FUNCNAME)                \
+  LogicalResult SimFHEEmitter::printOperation(OPTYPE op) {   \
+    std::string input = getName(op.GETTER());                \
+    os << "stats += evaluator." FUNCNAME "(" << input        \
+       << ", scheme_params.arch_param)\n";                   \
+    os << getName(op.getResult()) << " = " << input << "\n"; \
+    return success();                                        \
+  }
+
+// Since we only care about the cost, add/subtract are treated the same
```

Comment:
Is this implying that there is no `sub` in the SimFHE API? Better would be to make that explicit.

```suggestion
// SimFHE doesn't have `sub`. Since we only care about the cost it doesn't matter.
```